### PR TITLE
Show error dialogs to user when booting fails

### PR DESCRIPTION
**Pull request by MSuih**
This pull request adds dialog messages which inform the user when booting a game fails. This doesn't aim to give detailed techical explanations for the errors, but instead gives an simple, general explanation on the possible cause. The messages are also translateable.

As a side effect, makes emulator not log "Boot successful" when booting fails due to an error.

Example dialog:
image

Suggestions on how to improve wording on the messages are appriciated.

### CODE REVIEW

File: rpcs3/rpcs3qt/main_window.cpp [hunk 1]
```diff
@@ -2035,14 +2081,23 @@ void main_window::dropEvent(QDropEvent* event)
 		m_gameListFrame->Refresh(true);
 		break;
 	case drop_type::drop_game: // import valid games to gamelist (games.yaml)
-		if (Emu.BootGame(sstr(dropPaths.first()), "", true))
 		{
-			gui_log.success("Elf Boot from drag and drop done: %s", sstr(dropPaths.first()));
+			auto result = Emu.BootGame(sstr(dropPaths.first()), "", true);
+			if (result == game_boot_result::no_errors)
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/main_window.cpp]: switch this around and you can do
```suggestion
			if (const auto result = ... ; != game_boot_result::no_errors)
```

File: rpcs3/rpcs3qt/main_window.cpp [hunk 2]
```diff
@@ -1892,7 +1937,8 @@ void main_window::AddGamesFromDir(const QString& path)
 	{
 		std::string pth = sstr(dir_iter.next());
 
-		if (Emu.BootGame(pth, "", false, true))
+		auto result = Emu.BootGame(pth, "", false, true);
+		if (result == game_boot_result::no_errors)
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/main_window.cpp]:
```suggestion
		if (const auto result  = ...; result == game_boot_result::no_errors)
```

File: rpcs3/rpcs3qt/main_window.cpp [hunk 3]
```diff
@@ -1881,7 +1925,8 @@ void main_window::AddGamesFromDir(const QString& path)
 	const std::string s_path = sstr(path);
 
 	// search dropped path first or else the direct parent to an elf is wrongly skipped
-	if (Emu.BootGame(s_path, "", false, true))
+	Emu.BootGame(s_path, "", false, true);
+	if (true)
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/main_window.cpp]: ????

File: rpcs3/rpcs3qt/main_window.cpp [hunk 4]
```diff
@@ -261,7 +302,9 @@ void main_window::Boot(const std::string& path, const std::string& title_id, boo
 	Emu.SetForceBoot(true);
 	Emu.Stop();
 
-	if (Emu.BootGame(path, title_id, direct, add_only, force_global_config))
+	auto result = Emu.BootGame(path, title_id, direct, add_only, force_global_config);
+
+	if (result == game_boot_result::no_errors)
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/main_window.cpp]:
```suggestion
	if (const auto result =...; result != game_boot_result::no_errors)
```

File: rpcs3/rpcs3qt/main_window.cpp [hunk 5]
```diff
@@ -241,6 +241,47 @@ void main_window::OnPlayOrPause()
...
 }
 
+void main_window::show_boot_error(game_boot_result status)
+{
+	if (status == game_boot_result::no_errors)
+	{
+		return;
+	}
+	QString message;
+	switch (status) {
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/main_window.cpp]: I think our formatting puts case in the same tab position as switch.
Also:
```suggestion
	switch (status)
	{
```

File: rpcs3/rpcs3qt/main_window.cpp [hunk 6]
```diff
@@ -241,6 +241,47 @@ void main_window::OnPlayOrPause()
...
+		case game_boot_result::nothing_to_boot:
+			message = tr("No bootable content was found.");
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/main_window.cpp]: Break?

File: rpcs3/rpcs3qt/main_window.cpp [hunk 7]
```diff
@@ -241,6 +241,47 @@ void main_window::OnPlayOrPause()
...
+			break;
+		case game_boot_result::file_creation_error:
+			message = tr("Emulator could not create files required for booting.");
+			break;
+		case game_boot_result::generic_error:
+		default:
+			message = tr("Unknown error.");
+			break;
+	}
+	QString link = tr("<br /><br />For information on how to dump your PS3 games, read the <a href=\"https://rpcs3.net/quickstart\">quickstart guide</a>.");
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/main_window.cpp]: const

File: rpcs3/rpcs3qt/game_list_frame.cpp [hunk 8]
```diff
@@ -1238,17 +1238,18 @@ bool game_list_frame::CreatePPUCache(const game_info& game)
 	Emu.SetForceBoot(true);
 	Emu.Stop();
 	Emu.SetForceBoot(true);
-	const bool success = Emu.BootGame(game->info.path, game->info.serial, true);
+	const auto result = Emu.BootGame(game->info.path, game->info.serial, true);
 
-	if (success)
+	if (result == game_boot_result::no_errors)
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/game_list_frame.cpp]:
```suggestion
	if (const auto result =...; result != game_boot_result::no_errors)
```

File: rpcs3/Emu/System.cpp [hunk 9]
```diff
@@ -1617,7 +1639,10 @@ void Emulator::Stop(bool restart)
 	if (restart)
 	{
 		// Reload with prior configs.
-		return Load(m_title_id, false, m_force_global_config);
+		auto result = Load(m_title_id, false, m_force_global_config);
+		if (result != game_boot_result::no_errors)
```
Comment (Megamouse) [inline rpcs3/Emu/System.cpp]: const auto res
```suggestion
		if (const auto result =...; result != game_boot_result::no_errors)
```

File: rpcs3/Emu/System.cpp [hunk 10]
```diff
@@ -1572,7 +1591,10 @@ void Emulator::Stop(bool restart)
 		if (restart)
 		{
 			// Reload with prior configs.
-			return Load(m_title_id, false, m_force_global_config);
+			auto result = Load(m_title_id, false, m_force_global_config);
+			if (result != game_boot_result::no_errors)
```
Comment (Megamouse) [inline rpcs3/Emu/System.cpp]:
```suggestion
			if (const auto result =...; result != game_boot_result::no_errors)
```

File: rpcs3/Crypto/unself.cpp [hunk 11]
```diff
@@ -1322,9 +1322,7 @@ bool SELFDecrypter::GetKeyFromRap(u8* content_id, u8* npdrm_key)
 
 	if (!rap_file)
 	{
-		self_log.fatal("Failed to locate the game license file: %s."
-				  "\nEnsure the .rap license file is placed in the dev_hdd0/home/00000001/exdata folder with a lowercase file extension."
-				  "\nIf you need assistance on dumping the license file from your PS3, read our quickstart guide: https://rpcs3.net/quickstart", rap_path);
+		self_log.fatal("Failed to load RAP file: %s. Make sure the file extension is lowercase.", rap_path);
```
Comment (Megamouse) [inline rpcs3/Crypto/unself.cpp]: Why this weird revert

File: rpcs3/rpcs3qt/main_window.h [hunk 12]
```diff
@@ -17,6 +17,7 @@
 #include "gui_settings.h"
 #include "persistent_settings.h"
 #include "update_manager.h"
+#include "Emu/System.h"
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/main_window.h]: forward declaration of game_boot_result should suffice

**Comment by Megamouse** [comment]
You can skip the != noerror everywhere if no error is 0

**Comment by Megamouse** [comment]
Updated the things with something that might be more fitting

File: rpcs3/rpcs3qt/main_window.cpp (hunk 3 above)
Comment (MSuih) [reply rpcs3/rpcs3qt/main_window.cpp]: whoops, forgot to fill it in

File: rpcs3/Crypto/unself.cpp (hunk 11 above)
Comment (MSuih) [reply rpcs3/Crypto/unself.cpp]: The link is no longer needed as it's in the dialog box, and I dislike having such a long and lenghty message printed to log, I can drop the revert if you want to keep it around, but it's ugly and not neccessary.

**Comment by MSuih** [comment]
Fixed. I had to revert from class enum to plain enum to get boolean conversion to work.

File: rpcs3/rpcs3qt/game_list_frame.cpp [hunk 13]
```diff
@@ -1238,17 +1238,17 @@ bool game_list_frame::CreatePPUCache(const game_info& game)
...
-	const bool success = Emu.BootGame(game->info.path, game->info.serial, true);
 
-	if (success)
+	if (const auto error = Emu.BootGame(game->info.path, game->info.serial, true))
 	{
-		game_list_log.warning("Creating PPU Cache for %s", game->info.path);
+		game_list_log.error("Could not create PPU Cache for %s, error: %s", game->info.path, error);
+		return false;
 	}
 	else
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/game_list_frame.cpp]: I think you can rmove the else scope and simply return (I personally don't like that the return isn't the last thing in the outer scope)

File: rpcs3/rpcs3qt/main_window.cpp [hunk 14]
```diff
@@ -241,6 +241,49 @@ void main_window::OnPlayOrPause()
...
+		return;
+	}
+	QString message;
+	switch (status)
+	{
+	case game_boot_result::nothing_to_boot:
+		message = tr("No bootable content was found.");
+		break;
+	case game_boot_result::wrong_disc_location:
+		message = tr("Disc could not be mounted properly. Make sure the disc is not in dev_hdd0/game folder.");
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/main_window.cpp]: "Make sure the disc is not in dev_hdd0/game folder."

This sounds a bit wonky

File: rpcs3/rpcs3qt/main_window.cpp [hunk 15]
```diff
@@ -241,6 +241,49 @@ void main_window::OnPlayOrPause()
...
+		break;
+	case game_boot_result::invalid_file_or_folder:
+		message = tr("The selected file or folder is invalid or corrupted.");
+		break;
+	case game_boot_result::install_failed:
+		message = tr("Additional content could not be installed.");
+		break;
+	case game_boot_result::decryption_error:
+		message = tr("Digital content could not be decrypted. This is usually caused by missing or invalid license (RAP) file.");
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/main_window.cpp]: similarly wonky

File: rpcs3/rpcs3qt/main_window.cpp [hunk 16]
```diff
@@ -241,6 +241,49 @@ void main_window::OnPlayOrPause()
...
+		break;
+	case game_boot_result::file_creation_error:
+		message = tr("Emulator could not create files required for booting.");
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/main_window.cpp]:
```suggestion
		message = tr("The emulator could not create files required for booting.");
```

File: rpcs3/rpcs3qt/main_window.cpp [hunk 17]
```diff
@@ -2035,14 +2079,22 @@ void main_window::dropEvent(QDropEvent* event)
 		m_gameListFrame->Refresh(true);
 		break;
 	case drop_type::drop_game: // import valid games to gamelist (games.yaml)
-		if (Emu.BootGame(sstr(dropPaths.first()), "", true))
 		{
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/main_window.cpp]: weird scope

File: rpcs3/rpcs3qt/game_list_frame.cpp [hunk 18]
```diff
@@ -1238,17 +1238,9 @@ bool game_list_frame::CreatePPUCache(const game_info& game)
...
 	Emu.SetForceBoot(true);
-	const bool success = Emu.BootGame(game->info.path, game->info.serial, true);
 
-	if (success)
-	{
-		game_list_log.warning("Creating PPU Cache for %s", game->info.path);
-	}
-	else
-	{
-		game_list_log.error("Could not create PPU Cache for %s", game->info.path);
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/game_list_frame.cpp]: where did that go?

File: rpcs3/rpcs3qt/main_window.cpp [hunk 19]
```diff
@@ -1881,7 +1925,7 @@ void main_window::AddGamesFromDir(const QString& path)
 	const std::string s_path = sstr(path);
 
 	// search dropped path first or else the direct parent to an elf is wrongly skipped
-	if (Emu.BootGame(s_path, "", false, true))
+	if (const auto error = Emu.BootGame(s_path, "", false, true); error != game_boot_result::no_errors)
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/main_window.cpp]:
```suggestion
	if (const auto error = Emu.BootGame(s_path, "", false, true))
```

File: rpcs3/rpcs3qt/main_window.cpp [hunk 20]
```diff
@@ -1892,7 +1936,7 @@ void main_window::AddGamesFromDir(const QString& path)
 	{
 		std::string pth = sstr(dir_iter.next());
 
-		if (Emu.BootGame(pth, "", false, true))
+		if (const auto error = Emu.BootGame(pth, "", false, true); error != game_boot_result::no_errors)
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/main_window.cpp]:
```suggestion
		if (const auto error = Emu.BootGame(pth, "", false, true))
```

File: rpcs3/Crypto/unself.cpp (hunk 11 above)
Comment (Megamouse) [reply rpcs3/Crypto/unself.cpp]: but you are removing information that was added by helpers in order to get less spam

**Comment by MSuih** [comment]
Dropped the commit which alters log message

File: rpcs3/Emu/System.cpp [hunk 21]
```diff
@@ -512,21 +532,17 @@ bool Emulator::BootGame(const std::string& path, const std::string& title_id, bo
...
 	{
 		elf = path + elf;
 
 		if (fs::is_file(elf))
 		{
 			m_path = elf;
-			Load(title_id, add_only, force_global_config);
-			success = true;
-			break;
+			return Load(title_id, add_only, force_global_config);
```
Comment (Megamouse) [inline rpcs3/Emu/System.cpp]: You are completely breaking the logic here.
BootGame is not supposed to return before reaching the endof the function.

File: rpcs3/Emu/System.cpp [hunk 22]
```diff
@@ -546,14 +562,12 @@ bool Emulator::BootGame(const std::string& path, const std::string& title_id, bo
 				if (fs::is_file(elf))
 				{
 					m_path = elf;
-					Load(title_id, add_only, force_global_config);
-					success = true;
+					return Load(title_id, add_only, force_global_config);
```
Comment (Megamouse) [inline rpcs3/Emu/System.cpp]: same here

File: rpcs3/Emu/System.cpp [hunk 23]
```diff
@@ -512,20 +532,20 @@ bool Emulator::BootGame(const std::string& path, const std::string& title_id, bo
...
 	{
 		m_path = path;
-		Load(title_id, add_only, force_global_config);
-		return true;
+		return Load(title_id, add_only, force_global_config);
 	}
 
-	bool success = false;
+	bool found_any = false;
+	game_boot_result last_status = generic_error;
```
Comment (Megamouse) [inline rpcs3/Emu/System.cpp]: found_any is unnecessary.
Set the initial error to nothing_to_boot.
```suggestion
	game_boot_result result = game_boot_result::nothing_to_boot;
```

File: rpcs3/Emu/System.cpp [hunk 24]
```diff
@@ -546,14 +566,20 @@ bool Emulator::BootGame(const std::string& path, const std::string& title_id, bo
 				if (fs::is_file(elf))
 				{
 					m_path = elf;
-					Load(title_id, add_only, force_global_config);
-					success = true;
+					found_any = true;
+					if (last_status) // keep no_errors if even one load was successful
```
Comment (Megamouse) [inline rpcs3/Emu/System.cpp]: we only update boot_error if it was an error already
```suggestion
					if (const auto err = Load(title_id, add_only, force_global_config); err != game_boot_result::no_errors)
					{
						result = err;
					}
```

File: rpcs3/Emu/System.cpp [hunk 25]
```diff
@@ -546,14 +566,20 @@ bool Emulator::BootGame(const std::string& path, const std::string& title_id, bo
...
 			}
 		}
 	}
-
-	return success;
+	if (found_any)
+	{
+		return last_status;
+	}
+	return game_boot_result::nothing_to_boot;
```
Comment (Megamouse) [inline rpcs3/Emu/System.cpp]:
```suggestion
	return result;
```

File: rpcs3/Emu/System.h [hunk 26]
```diff
@@ -16,6 +16,18 @@ enum class system_state
 	ready,
 };
 
+enum game_boot_result : u32
```
Comment (elad335) [inline rpcs3/Emu/System.h]:
```suggestion
enum class game_boot_result : u32
```

Comment (MSuih) [reply rpcs3/Emu/System.h]: I'd rather not, class enums can't be implictly converted to booleans which means rewriting all if statements in this PR.

**Comment by Megamouse** [comment]
upsies. needs another rebase

File: rpcs3/Emu/System.h (hunk 26 above)
Comment (elad335) [reply rpcs3/Emu/System.h]: then why not overloading operator bool for it?

Comment (elad335) [reply rpcs3/Emu/System.h]:
```constexpr operator bool(game_boot_result res)
{
    return res != game_boot_result{};
}
```

Comment (MSuih) [reply rpcs3/Emu/System.h]: Sadly not supported for class enums

Comment (Nekotekina) [reply rpcs3/Emu/System.h]: Could sacrifice lack of operator bool for enum class I think.

**Comment by Megamouse** [comment]
needs another rebase

File: rpcs3/Emu/System.h (hunk 26 above)
Comment (MSuih) [reply rpcs3/Emu/System.h]: On second though yeah, seems more sensible.

**Pull request merged**